lib/Vend/Tax/Avalara.pm

```diff
@@ -0,0 +1,1343 @@
+
+Maps to Avalara required field date. Normally default should be fine, but if
+you need to specify the transaction for an alternative date, this can be
+overridden.
+
+=item B<product_tax_code_field> - default: "product_tax_code"
+
+Field name in sku's source table for Avalara's product tax code for the
+product in question. Since the lookup uses the C<item_field()> routine, the
+table should be found in I<ProductFiles> directive. Field is
+products.product_tax_code in strap demo schema.
+
+List of Avalara supported product tax codes:
+
+https://taxcode.avatax.avalara.com/
+
+=item B<shipping_tax_code> - default: "FR"
+
+Value to apply to taxCode field for the shipping line item.
+
+=item B<handling_tax_code> - default: "OH010000"
+
+Value to apply to taxCode field for the handling line item, if B<handling> >=
+0.01
+
+=item B<load_line_items> - default: undef
+
+Boolean to set to true when sending transaction data to Avalara if you want
+the order data to use derived directly from the orderline table based on
+B<order_number>. See strap Job "send_tax_transaction" for an example usage.
+
+=item B<sent_field> - default: "tax_sent"
+
+Name of field in transactions table to update to indicate the result of the
+attempt to report the transaction. On error (defined as an HTTP response code
+in the 400 range), this field will be set to "-1" to indicate that manual
+intervention is required to assess the order's situation.
+
+=item B<sent_success_value> - default: 1
+
+Value to update B<sent_field> to on a successful transaction submission.
+
+=item B<table> - default: __TAXAVERAGES_TABLE__ || "tax_averages"
+
+Overrides Vend::Tax::table by adding secondary fallback to "tax_averages".
+
+=item B<nexus_address> - default: __NEXUS_ADDRESS__
+
+Location address used to determine jurisdiction of merchant.
+
+=item B<nexus_city> - default: __NEXUS_CITY__
+
+Location city used to determine jurisdiction of merchant.
+
+=item B<nexus_state> - default: __NEXUS_STATE__
+
+Location state used to determine jurisdiction of merchant.
+
+=item B<nexus_zip> - default: __NEXUS_ZIP__
+
+Location zip used to determine jurisdiction of merchant.
+
+=item B<nexus_country> - default: __NEXUS_COUNTRY__
+
+Location country used to determine jurisdiction of merchant.
+
+=item B<no_nexus_required> - default: undef
+
+When Perly true, allows the post_transaction() routine called from
+[send-tax-transaction] to create the tax transaction regardless of nexus for
+the order's shipping jurisdiction. Allows for accounting with all orders in
+Avalara instead of only those orders that happen to collect tax.
+
+=item B<invoice_number> = default: undef
+
+Invoice number applied to tax transaction created in [send-tax-transaction].
+Typically would be set to the Interchange order number (and is set accordingly
+in send_tax_transaction job).
+
+=back
+
+=cut
+
+### accessors
+{
+    my @acc = (
+        json => q{JSON->new->pretty->allow_nonref->canonical},
```

Comment:
I always pretty print JSON. I care far less about the size difference of the payload than I do about being able to get logs I can read.

Reply:
How big are these request bodies? It can make a huge difference to overall traffic when they're big, but if they're small maybe it doesn't matter much.

Reply:
They're relatively small, and they are relatively infrequent. The results for a particular configuration are cached, so it's a single request/response for each configuration.

Reply:
Here's some sample logging:

`none - - [05/March/2023:14:28:33 -0500] strap /cgi-bin/strap DEBUG: post_transaction() full request
> POST https://sandbox-rest.avatax.com/api/v2/transactions/create
> Cache-Control: no-cache
> Authorization: Basic Z2FyeS5ub3J0b25Ab2RpbmRldi5jb206MGRpbmQzdkRheXN0YXI=
> User-Agent: libwww-perl/6.68
> Content-Length: 804
> Content-Type: application/json
> 
> {
>    "addresses" : {
>       "shipFrom" : {
>          "city" : "Anywhere",
>          "country" : "US",
>          "line1" : "123 Main St.",
>          "postalCode" : "45056",
>          "region" : "OH"
>       },
>       "shipTo" : {
>          "city" : "Norman",
>          "country" : "US",
>          "line1" : null,
>          "postalCode" : "73069",
>          "region" : "OK"
>       }
>    },
>    "code" : "SURUTVABA0003",
>    "commit" : true,
>    "customerCode" : "NONE",
>    "date" : "2023-03-05",
>    "lines" : [
>       {
>          "amount" : "36.99",
>          "itemCode" : "os28085",
>          "number" : 0,
>          "quantity" : 1
>       },
>       {
>          "amount" : 5.35,
>          "itemCode" : "Shipping",
>          "number" : "-1",
>          "quantity" : 1,
>          "taxCode" : "FR"
>       }
>    ],
>    "type" : "SalesInvoice"
> }
> 
`

Reply:
Ok, I won't stand in the way here. I think request/response bodies should always use "compressed" JSON and we can use `jq` or similar to pretty it up when viewing, but don't mind me here.